documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,888 @@
+- There are 2 routing and IP mapping stages in the pipeline, so we can support either source and destination IP matching, as well as 2 layers of routing.
+
+```mermaid
+flowchart LR
+    Routing0[Routing 0]
+    Routing1[Routing 1]
+    Map0[IP Mapping 0]
+    Map1[IP Mapping 1]
+    TcpPortMap[TCP Port Mapping]
+    UdpPortMap[UDP Port Mapping]
+
+    Routing0 --> | lpmrouting - 1 | Routing1
+    Routing1 --> | maprouting - 0 | Map0
+    Map0 --> | maprouting - 1 | Map1
+    Map1 --> | portmaprouting | TcpPortMap
+    Map1 --> | portmaprouting | UdpPortMap
+```
+
+#### 5.9.3. Stage transitions
+
+##### 5.9.3.1. Stage transition routing type
+
+To transit between stages, we use `transition` field to specify the transition routing type.
+
+A transition routing type is a special type of [Routing Type](#572-routing-type):
+
+1. Each transition routing type can only have one single routing action.
+2. Only a special list of routing actions will take effect in the routing types, which are all related to stage transitions, as listed below.
+3. If `transition` field is not set, we consider the packet is reaching the end of the pipeline and start executing the final routing actions.
+
+| `action_type` value | Stage |
+| ---------------- | ----- |
+| `drop` | Drop the packet |
+| `trap` | Sending it to CPU |
+| `lpmrouting` | Skip to next routing stage |
+| `maprouting` | Skip to next mapping stage |
+| `portmaprouting` | Skip to next TCP or UDP Port Mapping stage |
+
+To help us being flexible, the transition routing actions can take a few parameters:
+
+- All transition routing actions other than `drop` and `trap`:
+  - `default_routing_type`: If no entry is found, use this routing type to route the packet. If `default_routing_type` is not set, the packet will be dropped by default.
+  - `stage_index`: If there are multiple stage of the same stage type, use this field to specify which stage to transit to. The index is starting from 0 and by default 0.
+- IP-based routing actions (`lpmrouting` and `maprouting`):
+  - `use_src_ip`: Use source IP in routing and mapping stages.
+  - `ip_mask`: IP mask to apply before matching the entries.
+
+Here is an example that shows how the routing stage entry looks like:
+
+```json
+// Routing stage entry:
+// When this entry is matched, we will move to VNET mapping stage, which executes maprouting action and jump to VNET mapping stage.
+"DASH_SAI_ROUTE_TABLE|123456789012|0|10.0.1.0/24": {
+    "transition": "vnetmap",
+    "vnet": "Vnet1"
+}
+
+"DASH_SAI_ROUTING_TYPE_TABLE|vnetmap": [
+    {
+        "action_type": "maprouting",
+        "stage_index": 0
+    }
+]
+```
+
+##### 5.9.3.2. Stage skipping
+
+To avoid loop shows up in the pipeline, the pipeline is designed to be forward only. When transition routing action is invoked, we can simply update the target stage in the metadata, then the transition behavior can simply be described by the code below:
```

Comment:
```suggestion
To avoid loops appearing in the pipeline, the pipeline is designed to be forward only. When transition routing action is invoked, we can simply update the target stage in the metadata, then the transition behavior can simply be described by the code below:
```